sycl/source/detail/graph_memory_pool.hpp

```diff
@@ -0,0 +1,209 @@
+  struct alloc_info {
+    // Allocation kind
+    usm::alloc Kind;
+    // Size of the allocation
+    size_t Size;
+    // Is currently mapped to physical memory
+    bool Mapped = false;
+    // Index into the array of physical memory
+    size_t PhysicalMemID = 0;
+    // Is the allocation read only
+    bool ReadOnly = false;
+    // Should the allocation be zero initialized during initial allocation
+    bool ZeroInit = false;
+  };
+
+public:
+  graph_mem_pool(const context &Context, const device &Device)
+      : MContext(Context), MDevice(Device) {}
+  ~graph_mem_pool() {
+    for (auto &[Ptr, AllocInfo] : MAllocations) {
+      // Unmap allocations if required before physical memory is released
+      // Physical mem is released when MPhysicalMem is cleared
+      if (AllocInfo.Mapped) {
+        unmap(Ptr, AllocInfo.Size, MContext);
+      }
+      // Free the VA range
+      free_virtual_mem(reinterpret_cast<uintptr_t>(Ptr), AllocInfo.Size,
+                       MContext);
+    }
+    MPhysicalMem.clear();
+  }
+
+  /// Memory pool cannot be copied
+  graph_mem_pool(graph_mem_pool &) = delete;
+
+  /// Get a pointer to a new allocation. For device allocations these are
+  /// virtual reservations which must be later mapped to allocated physical
+  /// memory before use by calling allocateAndMapAll()
+  /// @param Size Size of the allocation
+  /// @param AllocType Type of the allocation
+  /// @param MemPool Option memory pool from which allocations will not be made
+  /// from directly but properties may be respected.
+  /// @return A pointer to the start of the allocation
+  void *malloc(size_t &Size, usm::alloc AllocType,
+               const std::shared_ptr<memory_pool_impl> &MemPool = nullptr) {
+    void *Alloc = nullptr;
+    switch (AllocType) {
+    case usm::alloc::device: {
+
+      auto &CtxImpl = sycl::detail::getSyclObjImpl(MContext);
+      auto &Adapter = CtxImpl->getAdapter();
+
+      size_t Granularity = get_mem_granularity(MDevice, MContext);
+      uintptr_t StartPtr = 0;
+      size_t AlignedSize = alignByteSize(Size, Granularity);
+      // Do virtual reservation
+      Adapter->call<sycl::errc::runtime,
+                    sycl::detail::UrApiKind::urVirtualMemReserve>(
+          CtxImpl->getHandleRef(), reinterpret_cast<void *>(StartPtr),
+          AlignedSize, &Alloc);
+
+      alloc_info AllocInfo = {};
+      AllocInfo.Kind = AllocType;
+      AllocInfo.Size = AlignedSize;
+
+      // Collect relevant properties from memory pool
+      if (MemPool) {
+        auto PropList = MemPool->getPropList();
+        if (PropList.has_property<property::memory_pool::zero_init>()) {
+          AllocInfo.ZeroInit = true;
+        }
+        if (PropList.has_property<property::memory_pool::read_only>()) {
+          AllocInfo.ReadOnly = true;
+        }
+      }
+
+      MAllocations[Alloc] = AllocInfo;
+      // Modify Size parameter as caller may need to know the adjusted size
+      Size = AlignedSize;
```

Comment:
At a first glance it seemed a bit strange to me that we pass `Size` by reference and as a result preventing from being able to call `malloc` with something like `num * sizeof(int)`. Is this required? Can't the caller know the AlignedSize from AllocInfo somehow?

Reply:
This is only called by the async alloc entry points which would never call in such a way (they have no knowledge of element size). To be fair I could probably remove passing by reference, I had it this way because I thought we might need to store the correct (possibly modified to be aligned) size in the handler/command group for alloc nodes. But that is not required yet so this was just future proofing and could be removed until we actually need it.

We could potentially require the caller to first call some function to align the size themselves, but we can't really enforce that so it seems a little error prone to me. We could also potentially take the pointer by reference and return the actual allocation size, though I don't like that much better.

Reply:
Ok, if it's only called by async alloc entry points then that's fine with me.